### ISSUE TITLE
[HOLD for payment 2023-10-09] [HOLD for payment 2023-09-04] [HOLD 25758][$2000] App displays strong HTML tag in reply in thread first message if we keep it in edit mode and edit message in main report

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21527]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Open any report
3. Send any message with ** around it to make it bold eg: *hello*
4. Hover and click on reply in thread
5. Edit the first message of new thread of reply in thread and keep it in edit mode
6. Open main report and edit the message on which we had clicked reply in thread
7. Again hover and click on reply in thread and observe that message now has <strong></strong> HTML tag around it
## Expected Result:
App should not display HTML tags around bold messages
## Actual Result:
App display's strong HTML tag around message in reply in thread if the message is kept in edit and we edit that message in main thread
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.32-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/a76fb5eb-a5a7-432b-9a24-880d826962a0

https://github.com/Expensify/App/assets/43996225/c52e9c04-e295-482d-8eed-be4329d7dd8d

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1687259310889789

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ee1f1adfd50b483f</li>
        <li>Upwork Job ID: 1675932442345897984</li>
        <li>Last Price Increase: 2023-07-20</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Reviewer | 26113077</li>
<li>eh2077 | Contributor | 26113080</li>
<li>dhanashree-sawant | Reporter | 26113082</li>
</ul></li>
    </ul>
</details>

**Comment by dylanexpensify** [comment]
Closing, @kavimuru feel free to open if you can repro

**Comment by eh2077** [comment]
@kavimuru @dylanexpensify I can still reproduce it on prod, staging and dev. See below video FYI

https://github.com/Expensify/App/assets/117511920/7ddbe37d-a401-4809-bb55-172ecaec6ff4

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

App displays strong HTML tag in reply in thread first message if we keep it in edit mode and edit message in main report

### What is the root cause of that problem?

The root cause of this issue is that 

1. When editing the first message of thread, we save draft message(`action.message[0].html`) for key determined by  `reportID` of thread and `reportActionID` of parent report action, see

   https://github.com/Expensify/App/blob/2f6002fd32845905662a363915d907ee222b8e67/src/pages/home/report/ContextMenu/ContextMenuActions.js#L258

2. After editing and saving the changed message from main report action and going back to thread, the following condition

   https://github.com/Expensify/App/blob/2f6002fd32845905662a363915d907ee222b8e67/src/pages/home/report/ReportActionItemMessageEdit.js#L84

   is false because `action.message[0].html` has been changed and the `props.draftMessage` is still same as the html of original report action. So, the draft is

   https://github.com/Expensify/App/blob/2f6002fd32845905662a363915d907ee222b8e67/src/pages/home/report/ReportActionItemMessageEdit.js#L90

   which displayes html tag if the original `props.action.message[0].html` is not plain text. Note that `props.draftMessage` which is saved in Step 1 is read from [here](https://github.com/Expensify/App/blob/64b720c5b74c3001bf3f9eceac1f54511611c6b7/src/pages/home/report/ReportActionItem.js#L531-L532) through `reportID` of thread and `reportActionID` of parent report action.

### What changes do you think we should make in order to solve the problem?

To fix this issue, we should save and read `draftMessage` for the first message of a thread using  `reportID` of parent report and `reportActionID` of parent report action. The first message of thread already uses the same report action of parent report. So, we only need to fix the `reportID`.

To achieve it, we can 

1. Change the second parameter of method [saveReportActionDraft](https://github.com/Expensify/App/blob/6b4bddfbd1ca892dd6971de5aa66b367e59db07c/src/libs/actions/Report.js#L1072) from `reportActionID` to `reportAction` and use to method `ReportUtils.getOriginalReportID` to get the right reportID. The changed method `saveReportActionDraft` will be like

   ```js
   /**
    * Saves the draft for a comment report action. This will put the comment into "edit mode"
    *
    * @param {String} reportID
    * @param {Number} reportAction
    * @param {String} draftMessage
    */
   function saveReportActionDraft(reportID, reportAction, draftMessage) {
       const originalReportID = ReportUtils.getOriginalReportID(reportID, reportAction);
       Onyx.set(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${originalReportID}_${reportAction.reportActionID}`, draftMessage);
   }
   ```

2. Change all usages of `Report.saveReportActionDraft` to pass `reportAction` instead of `reportActionID`

3. Read draft message from correct `reportID` by changing this line

   https://github.com/Expensify/App/blob/4bd18483bda41eff673fbd0197ef077112acaa6b/src/pages/home/report/ReportActionItem.js#L523

   to

   ```js
   const originalReportID = ReportUtils.getOriginalReportID(props.report.reportID, props.action);
   const draftKey = `${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${originalReportID}_${props.action.reportActionID}`;
   ```

After fixing, the editing state of the first message of a thread and the message from parent report will be synced.

<details>
  <summary>Click to see demo video</summary>

  https://github.com/Expensify/App/assets/117511920/f27cec55-e0fe-4bbb-860b-c3f62643bfe2
</details>

### What alternative solutions did you explore? (Optional)

N/A

**Comment by kavimuru** [comment]
@dylanexpensify Issue is still  reproduced by Dhanashree and Eric

https://github.com/Expensify/App/assets/43996225/cbc55a2b-8f7f-4258-b759-7cf485147bc3


https://github.com/Expensify/App/assets/43996225/8901c112-f131-4f23-b1a0-bc3a99bad757

**Comment by dylanexpensify** [comment]
reviewing

**Comment by dylanexpensify** [comment]
Confirmed!

**Comment by dylanexpensify** [comment]
No proposals, doubling

**Comment by DrLoopFall** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Editing the message in the reply thread is converted to markdown when it's edited in the main report.

### What is the root cause of that problem?
When the main report is edited, this condition becomes false
https://github.com/Expensify/App/blob/2f6002fd32845905662a363915d907ee222b8e67/src/pages/home/report/ReportActionItemMessageEdit.js#L84-L87
which results in the draft message being not converted to markdown, because the current assumption is that the message is already in markdown.

### What changes do you think we should make in order to solve the problem?
We can convert the message to markdown and save the draft upon clicking the edit button.

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

### What is the root cause of that problem?

https://github.com/Expensify/App/blob/2607bdeeb211e2b8153b9ea6937064c27fd3d2c7/src/pages/home/report/ReportActionItemMessageEdit.js#L84-L90

draftMessage is set with the message's html value when edit button clicked. When the message is changed elsewhere, draftMessage is not the same as html and so not converted to markdown as you can see in the above code.

This is the root cause

### What changes do you think we should make in order to solve the problem?

We can save the draft message in markdown format consistently.
1. Change the below line
https://github.com/Expensify/App/blob/2607bdeeb211e2b8153b9ea6937064c27fd3d2c7/src/pages/home/report/ContextMenu/ContextMenuActions.js#L258

with
```
        const parser = new ExpensiMark();
        const editAction = () => Report.saveReportActionDraft(
            reportID,
            reportAction.reportActionID,
            _.isEmpty(draftMessage) ? parser.htmlToMarkdown(getActionText(reportAction)).trim() : ''
        );
```

2. Remove the markdown conversion from draft initialization in ReportActionItemMessageEdit

```
    const [draft, setDraft] = useState(Str.htmlDecode(props.draftMessage));
```

This works as expected.

<details>
    <summary>Result</summary>


https://github.com/Expensify/App/assets/126839717/af5a37cf-fa0f-4f75-a3a7-cc4149372c14


</details>

### What alternative solutions did you explore? (Optional)

**Comment by dylanexpensify** [comment]
Nice, can we get an update today @aimane-chnaif? 🙇‍♂️

**Comment by dylanexpensify** [comment]
Reviewing today!

**Comment by dylanexpensify** [comment]
@aimane-chnaif can you bring this up in slack channel to get more weigh in on options? TY sir!! 🙇‍♂️

**Comment by dylanexpensify** [comment]
@aimane-chnaif mind recapping where we landed on the convo? Next steps?

**Comment by dylanexpensify** [comment]
@aimane-chnaif any update?

**Comment by dylanexpensify** [comment]
@aimane-chnaif bump

**Comment by dylanexpensify** [comment]
bumping in slack

**Comment by situchan** [comment]
Lack of votes from [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1689867403367019) after 2 bumps, no one has strong opinion on how to move this forward.

> Don't sync anything, they are independent. So B keeps original draft message even if A message is updated and saved.

So I'd like to go with this one

**Comment by eh2077** [comment]
@situchan I think it makes more sense to sync the draft between thread and original report because they're original from the same report action.

>  Sync A and B always so if A is in edit mode, B should be in edit mode as well and draft message is also synced

Moreover, just like we sync the emojis reactions between thread and original report, it'll be consistent to sync the draft as well. 

cc @dylanexpensify

**Comment by situchan** [comment]
@dylanexpensify should we open 3rd discussion again on slack? First of all, what's your vote on those options?

<img width="1063" alt="Screenshot 2023-08-07 at 8 28 09 PM" src="https://github.com/Expensify/App/assets/108292595/ed14837c-6770-439e-8fc8-c3a126d5b7d9">

**Comment by s-alves10** [comment]
@situchan 

Will you tell me your thought about [my proposal](https://github.com/Expensify/App/issues/21527#issuecomment-1623541080)?
This is for 3rd option(sync nothing)

**Comment by situchan** [comment]
I can evaluate only after expected result is confirmed

**Comment by dylanexpensify** [comment]
Reviewing

**Comment by dylanexpensify** [comment]
Bumped thread. Tagging internal

**Comment by dylanexpensify** [comment]
@techievivek should we put this on hold while we decide our plan forward?

**Comment by techievivek** [comment]
I believe we've reached a consensus that if the solution doesn't look tricky, we want to sync both the message and the draft state. So @situchan should be able to start reviewing the proposal now.

**Comment by dylanexpensify** [comment]
Amazing, ty! @situchan let's get an update today! 🙇‍♂️

**Comment by situchan** [comment]
@eh2077's [proposal](https://github.com/Expensify/App/issues/21527#issuecomment-1614903333) looks good to me.
That's the only solution to meet the expected result we landed on:

> we want to sync both the message and the draft state

@techievivek you wanna be engineer here as you already have full context?

🎀 👀 🎀 C+ reviewed

**Comment by dylanexpensify** [comment]
Nice! @techievivek let us know!

**Comment by techievivek** [comment]
Yes, I can look into it. I hope that is fine with you, @luacmartins 🙏 .

**Comment by techievivek** [comment]
Proposal look good to me, and I am surprised to see how straightforward the solution is. 🎉

**Comment by situchan** [comment]
@techievivek this issue needs to have `External` label for Melvin to work

**Comment by techievivek** [comment]
Oh yes, I just realized that sorry for spmmaing 🤣

**Comment by eh2077** [comment]
@situchan The PR https://github.com/Expensify/App/pull/24544 is ready for review

**Comment by dylanexpensify** [comment]
Nice one!

**Comment by AndrewGable** [comment]
FYI this introduced a deploy blocker and I am going to revert this PR. There are some ideas on how to fix this here: https://github.com/Expensify/App/pull/24544

**Comment by situchan** [comment]
Good catch, though I am unable to reproduce as edit composer disappeared quickly. I had tested that case as well while reviewing PR and not found any issue.


https://github.com/Expensify/App/assets/108292595/4db9f9ce-91a4-4063-ad9d-89ed34b0baba

**Comment by AndrewGable** [comment]
You can see reproduction here: https://github.com/Expensify/App/issues/25690#issue-1861620119

It should only happen on staging and dev as this hasn't hit production yet.

**Comment by situchan** [comment]
Just tested main & staging. Maybe my device is fast enough so barely noticeable?

https://github.com/Expensify/App/assets/108292595/c4978b76-1b03-4a1d-ae84-57fed12d837b

**Comment by AndrewGable** [comment]
Yep, that is possible. Also maybe your chat/account does not have as much data, but it was noticeable on my account.

**Comment by situchan** [comment]
FYI: The solution was reverted due to https://github.com/Expensify/App/issues/25690 and now we should find new solution to avoid draft message briefly showing before delete.
In the meantime, ReportActionCompose which we'll be touching here is being huge refactored by #25758 and also announced to hold any non-urgent issues which affect ReportActionCompose - https://expensify.slack.com/archives/C02NK2DQWUX/p1692874524400529
@techievivek let's put this on hold for now

**Comment by techievivek** [comment]
Yeah, I agree to put this on HOLD while the compose refactor is going on. Also, @Pujan92 might have suggested a fix for the regression here: https://github.com/Expensify/App/issues/25690#issuecomment-1688974918, which we can explore once we are back in action.

**Comment by dylanexpensify** [comment]
perfect!

**Comment by situchan** [comment]
https://github.com/Expensify/App/pull/25758 is now deployed to production.
@eh2077 please update your proposal to cover delete case.

**Comment by eh2077** [comment]
### Proposal
[Updated](https://github.com/Expensify/App/issues/21527#issuecomment-1614903333) proposal to cover the delete case.

4. Add an effect to handle case of going back by browser back button and deleting action by saving empty draft. We can add following effect below this [line](https://github.com/Expensify/App/blob/dd71b38513990379ec8c8ea80531d9b4627b284c/src/pages/home/report/ReportActionItemMessageEdit.js#L113)
```js
    useEffect(() => {
        // We want to early return if the action is deleted or it's an initial draft
        if (ReportActionsUtils.isDeletedAction(props.action) || props.draftMessage === props.action.message[0].html) {
            return;
        }
        setDraft(Str.htmlDecode(props.draftMessage));
    }, [props.draftMessage, props.action.message]);
```

cc @situchan

**Comment by situchan** [comment]
@eh2077 please test your solution in android emulator which is very slow. and confirm that there's no message flickering anymore

**Comment by situchan** [comment]
@eh2077's [new proposal](https://github.com/Expensify/App/issues/21527#issuecomment-1614903333) looks good.
But we may still hold this issue for merge freeze.

**Comment by dylanexpensify** [comment]
@situchan shall we hold?

**Comment by situchan** [comment]
> @situchan shall we hold?

As merge freeze is over, we can continue.

@techievivek do you agree @eh2077 raising PR?

**Comment by techievivek** [comment]
Yes, we can move forward with the @eh2077 fix.

**Comment by situchan** [comment]
> Yes, we can move forward with the @eh2077 fix.

Thanks. @eh2077 please go ahead and raise PR

**Comment by eh2077** [comment]
I'm working on the PR and it'll be ready for review tmr

**Comment by eh2077** [comment]
@techievivek I found a new issue which is reproducible in production. When deleting the first chat of thread by saving empty msg, the editing mode is closed after the deleting dialog pops up, see


https://github.com/Expensify/App/assets/117511920/fa8ca742-6af2-4065-ab5d-9fb2a7dd11c0

I think this can be fixed by removing
https://github.com/Expensify/App/blob/1862aff7d5a4c677b3c7ea0424be1f6361dcc91a/src/pages/home/report/ReportActionItemMessageEdit.js#L261-L273

Should we fix it together in the PR?

**Comment by situchan** [comment]
@eh2077 that seems like recent regression. Did you find offending PR?
I think you can report this in slack channel and claim bug reporting bounty.

Btw, I will let @techievivek decide whether to fix here or not.

**Comment by eh2077** [comment]
@situchan It's from this PR https://github.com/Expensify/App/pull/23706. I reported it here https://expensify.slack.com/archives/C049HHMV9SM/p1694791088007549

**Comment by eh2077** [comment]
The bug gh issue is https://github.com/Expensify/App/issues/27605

**Comment by dylanexpensify** [comment]
@techievivek mind if we get an update today/tomorrow? 🙇‍♂️

**Comment by techievivek** [comment]
Sorry for the delay here. Looks like we have already created a separate issue for the above regression so we can just focus on original issue here. @eh2077 @situchan

**Comment by eh2077** [comment]
@dylanexpensify Sorry for the delay! I'll submit the PR today.

**Comment by eh2077** [comment]
@situchan Please kindly help to review the PR https://github.com/Expensify/App/pull/28154 when you get time, thanks!

**Comment by dylanexpensify** [comment]
Payment summary:
- Reporter: @dhanashree-sawant $250
- Contributor: @eh2077 $2,000
- Contributor+: @situchan $2,000

**Comment by dylanexpensify** [comment]
Payments sent! @situchan can we get BZ checklist done

**Comment by dylanexpensify** [comment]
yes please @situchan, just in that issue!
